perf/src/sigverify.rs

```diff
@@ -442,45 +445,75 @@ pub fn generate_offsets(
     )
 }
 
-//inplace shrink a batch of packets
-pub fn shrink_batches(batches: &mut Vec<PacketBatch>) {
-    let mut valid_batch_ix = 0;
-    let mut valid_packet_ix = 0;
-    let mut last_valid_batch = 0;
-    for batch_ix in 0..batches.len() {
-        for packet_ix in 0..batches[batch_ix].len() {
-            if batches[batch_ix][packet_ix].meta().discard() {
-                continue;
-            }
-            last_valid_batch = batch_ix.saturating_add(1);
-            let mut found_spot = false;
-            while valid_batch_ix < batch_ix && !found_spot {
-                while valid_packet_ix < batches[valid_batch_ix].len() {
-                    if batches[valid_batch_ix][valid_packet_ix].meta().discard() {
-                        batches[valid_batch_ix][valid_packet_ix] =
-                            batches[batch_ix][packet_ix].clone();
-                        batches[batch_ix][packet_ix].meta_mut().set_discard(true);
-                        last_valid_batch = valid_batch_ix.saturating_add(1);
-                        found_spot = true;
-                        break;
+fn split_batches(batches: Vec<PacketBatch>) -> (Vec<BytesPacketBatch>, Vec<PinnedPacketBatch>) {
```

Comment:
Is this function implementing https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.partition_in_place ? Saving a few allocations here may not a major optimization (I'm not sure how long the batches vec is expected to be), but if it is sufficiently long we could use impl from https://github.com/rust-lang/rust/pull/62278/files and then do split_at_mut to get two independent mutable slices with no allocations. Feel free to ignore this in favor of just removing shrinking and/or handling this in separate PR =)

Reply:
I was thinking about using `partition`/`partition_in_place`, but then I would end up with `&[PacketBatch]` slices and the `shrink_bytes_batches`/`shrink_pinned_batches` functions would need to do some ugly matching like `let left = PacketBatch::Bytes(b) else { panic!() }`. But I could do that if you and @alessandrod are fine with that.

At the end, I don't think that allocation matters much, these are vectors of vectors (so practically pointers of pointes). I can check today if that really affects the profiles (I don't think so).